system-addon/common/Reducers.jsm

```diff
@@ -60,7 +67,27 @@ function TopSites(prevState = INITIAL_STATE.TopSites, action) {
   }
 }
 
+function Search(prevState = INITIAL_STATE.Search, action) {
+  switch (action.type) {
+    case at.SEARCH_STATE_UPDATED:
+      if (!action.data) {
+        return prevState;
+      }
+      return Object.assign({}, prevState, {
+        currentEngine: {
+          name: action.data.currentEngine.name,
+          icon: action.data.currentEngine.iconBuffer
+        },
+        engines: action.data.engines.map(engine => ({
+          name: engine.name,
+          icon: engine.iconBuffer
```

Comment:
Sounds good. Doesn't look like this would save much:
```js
let {currentEngine, engines} = action.data;
return Object.assign({}, prevState, {
  currentEngine,
  engines
});
```